packages/core/src/resources/manager.ts

```diff
@@ -0,0 +1,182 @@
+        }
+        if (lastForwardSlashPart && forwardSlashParts.length > 1) {
+            return lastForwardSlashPart || uri;
+        }
+        if (uri.includes('\\')) {
+            const parts = uri.split('\\');
+            return parts[parts.length - 1] || uri;
+        }
+        return uri;
+    }
+
+    async list(): Promise<ResourceSet> {
+        const resources: ResourceSet = {};
+
+        try {
+            const mcpClients = this.mcpManager.getClients();
+            for (const [serverName, client] of mcpClients.entries()) {
+                try {
+                    const resourceUris = await client.listResources();
+                    logger.debug(`📁 Server '${serverName}' has ${resourceUris.length} resources`);
+                    for (const uri of resourceUris) {
+                        const qualifiedUri = `mcp:${serverName}:${uri}`;
+                        const metadata: ResourceMetadata = {
+                            uri: qualifiedUri,
+                            name: this.extractResourceName(uri),
+                            description: `Resource from MCP server: ${serverName}`,
+                            source: 'mcp',
+                            serverName,
+                            metadata: { originalUri: uri, serverName },
+                        };
+                        resources[qualifiedUri] = metadata;
+                    }
+                } catch (error) {
+                    logger.warn(
+                        `Failed to list resources from MCP server '${serverName}': ${error instanceof Error ? error.message : String(error)}`
+                    );
+                }
+            }
+        } catch (error) {
+            logger.error(
+                `Failed to get MCP resources: ${error instanceof Error ? error.message : String(error)}`
+            );
+        }
+
+        if (this.internalResourcesProvider) {
+            try {
+                const internalResources = await this.internalResourcesProvider.listResources();
+                for (const resource of internalResources) {
+                    const qualifiedUri = `internal:${resource.uri.replace(/^fs:\/\//, '')}`;
+                    resources[qualifiedUri] = {
+                        ...resource,
+                        uri: qualifiedUri,
+                        source: 'custom',
+                        description: `${resource.description || 'Internal resource'}`,
+                    };
+                }
+            } catch (error) {
+                logger.error(
+                    `Failed to get internal resources: ${error instanceof Error ? error.message : String(error)}`
+                );
+            }
+        }
+
+        const totalResources = Object.keys(resources).length;
+        const mcpCount = Object.keys(resources).filter((k) => k.startsWith('mcp:')).length;
+        const internalCount = Object.keys(resources).filter((k) =>
+            k.startsWith('internal:')
```

Comment:
nit: increment a counter in the above loops while we are adding the resources, instead of filtering through Object.keys each time which adds O(n) for each [O(3n) vs O(n)]

(minor but could become a problem if we have hundreds of resources through directories)